From 36917c44882409111924c5ff76e61db113293026 Mon Sep 17 00:00:00 2001
From: stheinan <stheinan@maxlinear.com>
Date: Wed, 25 Sep 2024 15:19:59 +0530
Subject: [PATCH] WLANRTSYS-80475 MLD Vaps are enabled even though configuring
 both Vaps with different parameters.

Issue:
MLD Vaps are created even when hidden and isolate parameter is given different values in config.

Fix:
Add check to not bringup MLD Vaps when hidden and isolate parameters for different
and to ignore maxassoc config when MLO is enabled.

FIXUP to WLANRTSYS-60843 MLO changes: AP MLD configuration
---
 hostapd/main.c   | 6 ++++++
 src/ap/mxl_mld.c | 8 +++++---
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/hostapd/main.c b/hostapd/main.c
index 0b1a5e1e6..5ac01bf03 100644
--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -321,6 +321,12 @@ setup_mld:
 #endif /* CONFIG_VENDOR_MXL */
 
 		iface->probe_resp_offloads = capa.probe_resp_offloads;
+
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+		if(hapd->conf->mxl_bss_conf.mlo_enable)
+			hapd->iconf->ap_max_num_sta = 255; //Max STA initialized when MLO is enabled
+#endif
+
 		hapd->iconf->ap_max_num_sta = MIN(capa.max_stations, hapd->iconf->ap_max_num_sta);
 		for (i = 0; i < hapd->iface->num_bss; i++) {
 			iface->conf->bss[i]->max_num_sta = MIN(capa.max_stations, iface->conf->bss[i]->max_num_sta);
diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 7d7fefd8f..7695779b4 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -196,9 +196,11 @@ bool mxl_hostapd_verify_mld_config(struct p_ap_mld *mld)
 			(link1->conf->wpa != link2->conf->wpa) ||
 			(link1->conf->wpa_pairwise != link2->conf->wpa_pairwise) ||
 			(link1->conf->rsn_pairwise != link2->conf->rsn_pairwise) ||
-			(os_strncmp((const char *)link1->conf->ssid.ssid, (const char *)link2->conf->ssid.ssid, MAX(link1->conf->ssid.ssid_len, link2->conf->ssid.ssid_len)))) {
-		wpa_printf(MSG_ERROR,"MLD mxl_hostapd_verify_mld_config failed");
-		return false;
+			(os_strncmp((const char *)link1->conf->ssid.ssid, (const char *)link2->conf->ssid.ssid, MAX(link1->conf->ssid.ssid_len, link2->conf->ssid.ssid_len))) ||
+			(link1->conf->isolate != link2->conf->isolate) ||
+			(link1->conf->ignore_broadcast_ssid != link2->conf->ignore_broadcast_ssid)) {
+				wpa_printf(MSG_ERROR,"MLD mxl_hostapd_verify_mld_config failed");
+				return false;
 	}
 
 	return true;
-- 
2.43.0

